src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -420,9 +430,33 @@ protected override void ApplyProjectReferenceAdded(ProjectId projectId, ProjectR
 
             var vsProject = (VSProject)project.Object;
             vsProject.References.AddProject(refProject);
+
+            var undoManager = TryGetUndoManager();
+            undoManager?.Add(new RemoveProjectReferenceUndoUnit(
+                this, projectId, projectReference.ProjectId));
+        }
+
+        private OleInterop.IOleUndoManager TryGetUndoManager()
+        {
+            var documentTrackingService = this.Services.GetService<IDocumentTrackingService>();
+            if (documentTrackingService != null)
+            {
+                var documentId = documentTrackingService.GetActiveDocument() ?? documentTrackingService.GetVisibleDocuments().FirstOrDefault();
```

Comment:
Seems like a nice candidate for a helper. Do we do this elsewhere in the code base?